src/manage_nsfs/nc_master_key_manager.js

```diff
@@ -147,36 +192,56 @@ class NCMasterKeysManager {
 
     /**
      * _create_master_keys_exec executes get master keys executable and retuns its data
-     * @returns {Promise<Object>}
+     * @returns {Promise<void>}
      */
     async _create_master_keys_exec(master_key) {
-        const master_keys_json = JSON.stringify({ active_master_key: master_key });
-        await os_util.spawn(config.NC_MASTER_KEYS_PUT_EXECUTABLE, [exec_key_suffix], { input: master_keys_json, stdio: [] });
+        const master_keys_json = this._get_master_keys_json(master_key);
+        try {
+            const put_master_keys_res = await os_util.spawn(config.NC_MASTER_KEYS_PUT_EXECUTABLE, [exec_key_suffix],
+                { shell: '/bin/sh', input: master_keys_json, stdio: [], return_stdout: true });
+            const { status } = JSON.parse(put_master_keys_res);
+
+            if (status === EXEC_STATUS_MAP.ok) {
+                dbg.log0('_create_master_keys_exec: finished successfully');
+            } else if (status === EXEC_STATUS_MAP.version_already_exist) {
+                dbg.warn(`_create_master_keys_exec: master keys file already exist status=${status}`);
+            } else {
+                throw new Error(`_create_master_keys_exec: failed with error=${status}`);
+            }
+        } catch (err) {
+            dbg.warn('_create_master_keys_exec: failed with error: ', err);
+            throw err;
+        }
     }
 
     /**
      * _init_from_exec init master keys from executable
+     * @returns {Promise<void>}
      */
     async _init_from_exec() {
-        const command = `${config.NC_MASTER_KEYS_GET_EXECUTABLE} ${exec_key_suffix}`;
-        for (let retries = 0; retries < 3;) {
+        const init_version = 0;
+        const command = `${config.NC_MASTER_KEYS_GET_EXECUTABLE} ${exec_key_suffix} ${init_version}`;
+        for (let retries = 0; retries < config.MASTER_KEYS_EXEC_MAX_RETRIES;) {
             try {
                 if (this.last_init_time &&
                     (new Date()).getTime() - this.last_init_time > config.NC_MASTER_KEYS_MANAGER_REFRESH_THRESHOLD) return;
-                const master_keys = await os_util.exec(command, { return_stdout: true });
-                const parsed_master_keys = JSON.parse(master_keys);
-                this._set_keys(parsed_master_keys);
-                this.last_init_time = (new Date()).getTime();
-                return;
-            } catch (err) {
-                if (err.stderr && err.stderr.trim() === 'NO_SUCH_KEY') {
-                    dbg.warn('init_from_exec: couldn\'t find master keys', err, err.code);
+                const get_master_keys_res = await os_util.exec(command, { return_stdout: true });
+                const { status, version, data } = JSON.parse(get_master_keys_res);
```

Comment:
We are reloading from executable every 5 minutes, and I worry that this behavior is error prone and could block access.
Maybe we should keep this config option but set the default to -1 in order to disable automatic refresh, WDYT?

Reply:
np, changed to -1